### DISCUSSION TITLE
Как организовать таблицу товаров с большим набором полей?

### DISCUSSION THREAD

**Discussion by OAndrey1985** [totumonline/totum-ru-issues-and-discussions#483] [Как это сделать?] [answered]
Начал разбираться с системой, очень крутая как по мне, но есть пара нюансов, как организовать таблицу в которой много полей. Ну допустим товарная позиция у которой название, артикул, количество, 10 характеристик, цвет, размер, вес, ширина, длина, толщина, тип, пол и т.д. товаров будет со временем более 30 тыс, + еще будут в зависимости от характеристик и опции, то есть условно товар делится как бы на блоки (основная информация, название, код, цена и т.д. фото/характеристки/опции/доп информация)  - это превый вопрос.
Вопрос 2. Привязка к категориям, создал справочник категорий и в товарной таблице селект, может лучше сделать как то таблицу циклов, чтобы проваливалась во внутрь?
Вопрос 3. Как посмотреть строку в отдельном окне, чтобы в popup например было выпадающее поле с набором полей разделенные на блоки или вкладки, вот например чтобы в таблице выводить базовые поля, а все эти характеристики и прочее запихнуть во фрейм например.
Понимаю что можно скрыть поля и логично запихнуть в одну таблицу все эти поля, но хотелось бы спросить тех у кого больше опыта, чтобы потом не переделывать.
Заранее спасибо.

**Comment by erpic-ru** [comment, marked as answer]
1. Делаете в таблице циклов, а товар - расчетная в цикле с кучей полей и вкладок
2. Делаете ссылку на категорию, либо [дерево](https://ru.docs.totum.online/treeview-based-on-select-tree)
3. linktopanel или опять же циклы. Ещё люди делают через временные таблицы

[Стоит пройти учебный курс](https://ru.docs.totum.online/training-course)

**Reply by erpic-ru** [reply]
п. 2 Ещё есть вариант [с ярлыками в меню](https://ru.docs.totum.online/system-tables#stables-tree) с открытием таблицы с префильтром

**Reply by ttm-alexeypolunin** [reply]
Еще вот ниже можно посмотреть: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/483#discussioncomment-7791021

**Comment by OAndrey1985** [comment]
> 1. Делаете в таблице циклов, а товар - расчетная в цикле с кучей полей и вкладок

Самый логичный способ, но в инструкции сказано, что расчетные таблицы и таблицы циклов делать большими массивами не правильно, если будет там 30 000 позиций или 100 тыс позиций, это не повлияет на работоспособность? Хотя я понимаю что в товаре в принципе расчетов никаких не происходит, только при добавлении, однако они же участвуют в заказах и так далее, еще планирую производство подключить с циклом процесса и расчета оплат сырья и так далее.

**Reply by erpic-ru** [reply]
У вас же товаров будет 30000 - записей циклов, а не 30000 строк и 30000 полей внутри цикла. Циклы - простая таблица, там хоть миллион может быть

**Comment by OAndrey1985** [comment]
> п. 2 Ещё есть вариант [с ярлыками в меню](https://ru.docs.totum.online/system-tables#stables-tree) с открытием таблицы с префильтром

Дерево логичнее, как то пропустил это в уроке, спасибо

**Comment by OAndrey1985** [comment]
> У вас же товаров будет 30000 - записей циклов, а не 30000 строк и 30000 полей внутри цикла. Циклы - простая таблица, там хоть миллион может быть

Извините за тупые вопросы, немного еще не понял до конца про таблицы циклов. Но разве 1 товар (с полями и значениями характеристиками, ценами и так далее) это не одна строка? То есть 30 тыс товаров 30 строк, просто они разбиты на расчетные таблицы которые так сказать собираются в таблицу цикла, и в этих расчетных таблицах будут строки характеристик для каждого товара (30 тыс), опций для каждого товара (30 тыс) и так далее, не повесит это систему? 
Или мне просто надо еще раз посмотреть и вникнуть в таблицу циклов? )

**Reply by ttm-alexeypolunin** [reply]
Таблица Циклов — это расширенная Простая таблица. Она физическая таблица в БД. В ней может быть сотня миллионов строк (вам потребуется сервер помощнее, чем обычно).

30К товаров — это фигня. Больше 3М — потребуется сделать индексы (в настройках таблицы) и подкрутить настройки постгреса (например 128Мб памяти у PG по умолчанию маловато на таблицы с 3М строк и больше).

Все таблицы внутри цикла (расчетные в цикле) — это виртуальные таблицы, каждая из которых хранится в одном поле БД и поэтому они и считаются всегда целиком. Именно поэтому их нельзя сделать большими.

Таблицы внутри цикла отделены от строки в таблице циклов. Когда изменяется строка в табл Циклов, ничего внутри цикла автоматически не перечитывается (надо спец код действия писать для этого). Обратно пересчитывается. При изменении внутри цикла в самом конце один раз пересчитывается строка в таблице циклов.

Поэтому для Заказов например с которыми работает 100 человек одновременно таблица циклов ставиться Без отслеживания изменений, что бы пользователи работающие в разных циклах не сталкивались по одновременному доступу.

В вашем случае, всякие настройки параметров товара отлично засовываются в таблицу внутрь цикла. Там же и делаете логику, что если товар такой, то такие поля, если другой то другие показывать и все, что вы там хотите наворотить.

**Reply by ttm-alexeypolunin** [reply]
Вопрос 3. — для циклов если не хотите проваливание, модифицируется стандартный код кнопки Открыть. Там по умолчанию self — замените на iframe и укажите размеры окна.

**Reply by OAndrey1985** [reply]
ага, понял, буду пробовать

**Reply by ttm-alexeypolunin** [reply]
А еще можно сделать кнопки в панели по правой кнопке мыши на поле с кнопкой открыть. Так тоже часто используется — стандарт открывает с проваливанием, а по правой кнопке можно открыть в окне.

Смотрите ф. `panelButton` в доках.
